pkg/agent/apiserver/handlers/podinterface/handler.go

```diff
@@ -0,0 +1,92 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podinterface
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/interfacestore"
+	"github.com/vmware-tanzu/antrea/pkg/monitor"
+)
+
+// Response describes the internal response struct of the pod command.
+type Response struct {
+	PodName       string `json:"name,omitempty" yaml:"name,omitempty" antctl:"name,The name of the pod"`
+	InterfaceName string `json:"InterfaceName,omitempty" yaml:"InterfaceName,omitempty"`
+	IP            string `json:"IP,omitempty" yaml:"IP,omitempty"`
+	MAC           string `json:"MAC,omitempty" yaml:"MAC,omitempty"`
+	PortUUID      string `json:"PortUUID,omitempty" yaml:"PortUUID,omitempty"`
+	OFPort        int32  `json:"OFPort,omitempty" yaml:"OFPort,omitempty"`
+	ContainerID   string `json:"ContainerID,omitempty" yaml:"ContainerID,omitempty"`
+	PodNamespace  string `json:"PodNamespace,omitempty" yaml:"PodNamespace,omitempty"`
+}
+
+func generateResponse(podName string, i *interfacestore.InterfaceConfig) Response {
+	return Response{
+		PodName:       podName,
+		InterfaceName: i.InterfaceName,
+		IP:            i.IP.String(),
+		MAC:           i.MAC.String(),
+		PortUUID:      i.OVSPortConfig.PortUUID,
+		OFPort:        i.OVSPortConfig.OFPort,
+		ContainerID:   i.ContainerInterfaceConfig.ContainerID,
+		PodNamespace:  i.ContainerInterfaceConfig.PodNamespace,
+	}
+}
+
+// HandleFunc returns the function which can handle queries issued by the pod command,
+func HandleFunc(aq monitor.AgentQuerier) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		name := r.URL.Query().Get("name")
+		ns := r.URL.Query().Get("namespace")
+		allNamespaces := r.URL.Query().Get("all-namespaces")
+
+		if len(name) > 0 && allNamespaces == "true" {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		var pods []Response
+		for _, v := range aq.GetInterfaceStore().GetInterfacesByType(interfacestore.ContainerInterface) {
+			podName := (*v.ContainerInterfaceConfig).PodName
+			podNS := (*v.ContainerInterfaceConfig).PodNamespace
+
+			if ns == podNS && podName == name {
+				pod := generateResponse(podName, v)
+				err := json.NewEncoder(w).Encode(pod)
+				if err != nil {
+					w.WriteHeader(http.StatusInternalServerError)
+				}
+				return
+			}
+
+			if allNamespaces == "true" || ns == podNS {
+				pod := generateResponse(podName, v)
+				pods = append(pods, pod)
+			}
```

Comment:
The logic seems wrong, all pods in a namespace will be returned if the specified pod is not found.
Can we have unit tests for `HandleFunc`?

Reply:
I think the logic is ok because of the `if` statement below:
```
if len(name) > 0 {
			w.WriteHeader(http.StatusNotFound)
}
```
but the code was also a bit confusing to me at first.

would definitely like to see some unit tests for the different cases

Reply:
Unit tests added.